### PR TITLE
Docs: Update linkcode implementation

### PR DESCRIPTION
**Pull request by kmaehashi**
Unwrap functions wrapped by `contextmanager` to fix `[source]` deadlink, e.g.:

* https://docs.cupy.dev/en/v10.0.0rc1/reference/generated/cupy.cuda.using_allocator.html?highlight=[source]
* https://docs.cupy.dev/en/v10.0.0rc1/reference/generated/cupyx.optimizing.optimize.html?highlight=[source]

Add `[source]` link to objects defined in C-extensions, e.g.:

* https://docs.cupy.dev/en/v10.0.0rc1/reference/generated/cupy.cuda.Device.html

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -434,6 +435,14 @@ def linkcode_resolve(domain, info):
     if not mod.__name__.split('.')[0] in _top_modules:
         return None
 
+    # If it's wrapped (e.g., by `contextlib.contextmanager`), unwrap it
+    for _ in range(10):
+        if not hasattr(obj, '__wrapped__'):
+            break
+        obj = obj.__wrapped__
+    else:
+        raise RuntimeError(f'nested too deep: {info}')
```
Comment (toslunar) [inline docs/source/conf.py:444]:
```suggestion
    obj = inspect.unwrap(obj)
```

File: docs/source/conf.py [hunk 2]
```diff
@@ -443,17 +452,24 @@ def linkcode_resolve(domain, info):
...
+        _, linenum = inspect.getsourcelines(obj)
+        assert isinstance(linenum, int)
 
     filename = os.path.realpath(filename)
     relpath = _get_source_relative_path(filename)
 
-    return 'https://github.com/cupy/cupy/blob/{}/{}#L{}'.format(
-        tag, relpath, linenum)
+    return 'https://github.com/cupy/cupy/blob/{}/{}'.format(
+        tag, relpath if linenum is None else f'{relpath}#L{linenum}')
```
Comment (toslunar) [inline docs/source/conf.py:472]: Could you avoid repeating `relpath`?  For example,
```python
fragment = ''
...
    fragment = f'#L{linenum}'
...
return f'https://github.com/cupy/cupy/blob/{tag}/{relpath}{fragment}'

**Review by toslunar — CHANGES_REQUESTED** [review]

File: docs/source/conf.py (hunk 1 above)
Comment (kmaehashi) [reply docs/source/conf.py]: wow didn't know this, thanks :)

**Review by toslunar — APPROVED** [review]
LGTM

**Pull request merged**